text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,648 @@
+   * @default - a generated name
+   */
+  readonly originAccessControlName?: string;
+  /**
+   * The type of origin that this origin access control is for.
+   * @default s3
+   */
+  readonly originAccessControlOriginType?: OriginAccessControlOriginType;
+  /**
+   * Specifies which requests CloudFront signs.
+   * @default always
+   */
+  readonly signingBehavior?: SigningBehavior;
+  /**
+   * The signing protocol of the origin access control.
+   * @default sigv4
+   */
+  readonly signingProtocol?: SigningProtocol;
+}
+
+/**
+ * Origin types supported by origin access control.
+ */
+export enum OriginAccessControlOriginType {
+  /**
+   * Uses an Amazon S3 bucket origin.
+   */
+  S3 = 's3',
+  /**
+   * Uses an AWS Elemental MediaStore origin.
+   */
+  MEDIASTORE = 'mediastore',
+  /**
+   * Uses a Lambda function URL origin.
+   */
+  LAMBDA = 'lambda',
+  /**
+   * Uses an AWS Elemental MediaPackage v2 origin.
+   */
+  MEDIAPACKAGEV2 = 'mediapackagev2',
+}
+
+/**
+ * Options for which requests CloudFront signs.
+ * Specify `always` for the most common use case.
+ */
+export enum SigningBehavior {
+  /**
+   * Sign all origin requests, overwriting the Authorization header
+   * from the viewer request if one exists.
+   */
+  ALWAYS = 'always',
+  /**
+   * Do not sign any origin requests.
+   * This value turns off origin access control for all origins in all
+   * distributions that use this origin access control.
+   */
+  NEVER = 'never',
+  /**
+   * Sign origin requests only if the viewer request
+   * doesn't contain the Authorization header.
+   */
+  NO_OVERRIDE = 'no-override',
+}
+
+/**
+ * The signing protocol of the origin access control.
+ */
+export enum SigningProtocol {
+  /**
+   * The AWS Signature Version 4 signing protocol.
+   */
+  SIGV4 = 'sigv4',
+}
+
+/**
+ * An Origin Access Control.
+ * @resource AWS::CloudFront::OriginAccessControl
+ * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudfront-originaccesscontrol.html
+ */
+export class OriginAccessControl extends OriginAccessControlBase {
+  /**
+   * Imports an origin access control from its id.
+   */
+  public static fromOriginAccessControlId(scope: Construct, id: string, originAccessControlId: string): IOriginAccessControl {
+    class Import extends OriginAccessControlBase {
+      public readonly originAccessControlId = originAccessControlId;
+      constructor(s: Construct, i: string) {
+        super(s, i);
+
+        this.originAccessControlId = originAccessControlId;
+      }
+    }
+    return new Import(scope, id);
+  }
+
+  public readonly originAccessControlId: string;
+  constructor(scope: Construct, id: string, props: OriginAccessControlProps = {}) {
+    super(scope, id);
+
+    const resource = new CfnOriginAccessControl(this, 'Resource', {
+      originAccessControlConfig: {
+        description: props.description,
+        name: props.originAccessControlName ?? this.generateName(),
```

Comment:
Rather than using `this.generateName()` we have logic that will do this for you. In the `super` call, you would do the following:
```ts
super(scope, id, { physicalName: props.originaAccessControlName ?? PhysicalName.GENERATE_IF_NEEDED });
```

Then, the name property here can be set using `physicalName`:
```ts
name: this.physicalName
```

Reply:
Double check me on this, but I'm pretty sure this is the way to go.

Reply:
I tried using `PhysicalName.GENERATE_IF_NEEDED` as suggested but it only generates a name if the resource is referenced across environments. The `name` property is mandatory so this fails during synth. Discovered the `Names.uniqueResourceName()` function though so no need to write our own name generating function :)